gammapy/estimators/map_profile.py

```diff
@@ -0,0 +1,313 @@
+                                        1.*u.deg,
+                                        11)
+
+        # set up profile estimator and run
+        prof_maker = MapProfileEstimator(boxes, axis)
+        fermi_prof = prof_maker.run(datasets[0])
+
+        # plot directly the data from the output dictionnary
+        x=fermi_prof.table["x_ref"]
+        y=fermi_prof.table["excess"]
+        yerr = [-fermi_prof.table["errn"], fermi_prof.table["errp"]]
+        plt.errorbar(x,y,yerr=yerr, fmt='o')
+        plt.yscale('log')
+        plt.show()
+
+        # smooth and plot the data using the ImageProfile class
+        fermi_prof.peek()
+        plt.show()
+
+        ax = plt.gca()
+        ax.set_yscale('log')
+        ax = fermi_prof.plot("flux", ax=ax)
+
+    """
+
+    def __init__(self, regions, axis, n_sigma = 1, n_sigma_ul=3):
+        self.regions = regions
+        self.axis = axis
+        self.n_sigma = n_sigma
+        self.n_sigma_ul = n_sigma_ul
+
+    def make_prof(self, dataset, steps):
+        """ Utility to make the profile in each region
+
+        Parameters
+        ----------
+        dataset : `~gammapy.datasets.MapDataset` or `~gammapy.datasets.MapDatasetOnOff`
+            the dataset to use for profile extraction
+        steps : list of str
+            the steps to be used.
+        Returns
+        --------
+        results : list of dictionnary
+            the list of results (list of keys: x_min, x_ref, x_max, alpha, counts, background, excess, ts, sqrt_ts, err,
+             errn, errp, ul, exposure, solid_angle)
+        """
+        if steps == "all":
+            steps = ["err", "ts", "errn-errp", "ul"]
+
+        results = []
+        for index, reg in enumerate(self.regions):
+            spds = dataset.to_spectrum_dataset(reg)
+            mask = spds.mask if spds.mask is not None else slice(None)
+            if isinstance(spds, SpectrumDatasetOnOff):
+                stats = WStatCountsStatistic(
+                    spds.counts.data[mask].sum(),
+                    spds.counts_off.data[mask].sum(),
+                    spds.alpha.data[0,0,0] # At some point, should replace with averaging over energy
+                )
+            else:
+                stats = CashCountsStatistic(
+                    spds.counts.data[mask].sum(),
+                    spds.background.data[mask].sum(),
+                )
+
+            result = {
+                "x_min": self.axis.edges[index],
+                "x_max": self.axis.edges[index+1],
+                "x_ref": self.axis.center[index]
+            }
+            if isinstance(spds, SpectrumDatasetOnOff):
+                result["alpha"] = stats.alpha
+            result.update({
+                "counts": stats.n_on,
+                "background": stats.background,
+                "excess": stats.excess,
+            })
+
+            if "ts" in steps:
+                if isinstance(stats.delta_ts, list) or isinstance(stats.delta_ts, np.ndarray):
```

Comment:
This type-casting and instance checks (all below) should not be needed...please remove.

Reply:
There is a problem of difference of type in the return values between WStatCountsStatistic and CashCountsStatistic. This was my trick to unify the outputs and to make work all other functions...

Reply:
Typically all functions in gammapy should just preserve the input type, unless there is a very good reason (e.g. performance, precision or memory consumption). That is what numpy does by default as well.

What is the exact problem that you observed in the `CashCountsStatistics` class? We should definitely try to fix it there, otherwise we will repeat the same code in all places where `CashCountsStatistics` is used...

Reply:
For one case, one returns an array, the other a float. This leads afterwards issue when creating the final table, propagating this issue up to the plotting function for the ULs... (ie the line 295 and 302 of the last version of image_profile.py)

Reply:
Fixed and thus simply, thanks to a PR of Régis